api_tests/users/views/test_user_actions.py

```diff
@@ -0,0 +1,229 @@
+        # Unauthorized user can't accept
+        res = app.post_json_api(url, accept_payload, expect_errors=True)
+        assert res.status_code == 401
+
+        # A random user can't accept
+        res = app.post_json_api(url, accept_payload, auth=some_rando.auth, expect_errors=True)
+        assert res.status_code == 403
+
+        # Moderator from another provider can't accept
+        another_moderator = AuthUserFactory()
+        another_moderator.groups.add(GroupHelper(PreprintProviderFactory()).get_group('moderator'))
+        res = app.post_json_api(url, accept_payload, auth=another_moderator.auth, expect_errors=True)
+        assert res.status_code == 403
+
+        # Node admin can't accept
+        res = app.post_json_api(url, accept_payload, auth=node_admin.auth, expect_errors=True)
+        assert res.status_code == 403
+
+        # Still unchanged after all those tries
+        preprint.refresh_from_db()
+        assert preprint.reviews_state == 'pending'
+        assert not preprint.is_published
+
+        # Moderator can accept
+        res = app.post_json_api(url, accept_payload, auth=moderator.auth)
+        assert res.status_code == 201
+        preprint.refresh_from_db()
+        assert preprint.reviews_state == 'accepted'
+        assert preprint.is_published
+
+    def test_cannot_create_actions_for_unmoderated_provider(self, app, url, preprint, provider, node_admin):
+        provider.reviews_workflow = None
+        provider.save()
+        submit_payload = self.create_payload(preprint._id, trigger='submit')
+        res = app.post_json_api(url, submit_payload, auth=node_admin.auth, expect_errors=True)
+        assert res.status_code == 409
+
+    def test_bad_requests(self, app, url, preprint, provider, moderator):
+        invalid_transitions = {
+            'post-moderation': [
+                ('accepted', 'accept'),
+                ('accepted', 'submit'),
+                ('initial', 'accept'),
+                ('initial', 'edit_comment'),
+                ('initial', 'reject'),
+                ('pending', 'submit'),
+                ('rejected', 'reject'),
+                ('rejected', 'submit'),
+            ],
+            'pre-moderation': [
+                ('accepted', 'accept'),
+                ('accepted', 'submit'),
+                ('initial', 'accept'),
+                ('initial', 'edit_comment'),
+                ('initial', 'reject'),
+                ('rejected', 'reject'),
+            ]
+        }
+        for workflow, transitions in invalid_transitions.items():
+            provider.reviews_workflow = workflow
+            provider.save()
+            for state, trigger in transitions:
+                preprint.reviews_state = state
+                preprint.save()
+                bad_payload = self.create_payload(preprint._id, trigger=trigger)
+                res = app.post_json_api(url, bad_payload, auth=moderator.auth, expect_errors=True)
+                assert res.status_code == 400
```

Comment:
When you fix the item from the top, this will turn into a 409. The request formatting is valid, it just conflicts with an existing database state.